File: daybreak/src/daybreak/core/orchestrator.py
import logging

from daybreak.config import config
from daybreak.core.artifacts import generate_artifacts
from .theme_model import normalize_mode
from .theme_registry import ThemeRegistry

logger = logging.getLogger("daybreak")


class ThemeOrchestrator:
    def __init__(self, system_adapter=None, terminal_adapters=None, registry=None):
        self.system_adapter = system_adapter
        self.terminal_adapters = terminal_adapters or []
        self.registry = registry or ThemeRegistry()

    def get_current_mode(self) -> str:
        if not self.system_adapter:
            return "light"
        try:
            return normalize_mode(self.system_adapter.get_current_mode())
        except Exception as exc:
            logger.warning(f"Failed to detect current mode. Defaulting to light: {exc}")
            return "light"

    def apply(self, mode: str, theme_name: str = None) -> str:
        normalized_mode = normalize_mode(mode)
        resolved_theme = theme_name or config.get_mode_theme_name(normalized_mode)
        palette = self.registry.get_palette(resolved_theme, normalized_mode)

        if self.system_adapter:
            self.system_adapter.set_mode(normalized_mode, palette=palette)

        for adapter in self.terminal_adapters:
            try:
                adapter.apply_mode(normalized_mode, resolved_theme, palette)
            except Exception as exc:
                name = getattr(adapter, "name", adapter.__class__.__name__)
                logger.warning(f"Terminal adapter '{name}' failed: {exc}")

        try:
            tokens = self.registry.get_tokens(resolved_theme, normalized_mode)
            accent_tokens = self.registry.get_accent_tokens(resolved_theme, normalized_mode)
            generate_artifacts(config.config_dir, resolved_theme, normalized_mode, tokens, accent_tokens, palette)
        except Exception as exc:
            logger.warning(f"Failed to generate theme artifacts: {exc}")

        return resolved_theme

    def apply_toggle(self, explicit_theme_name: str = None):
        target_mode = "light" if self.get_current_mode() == "dark" else "dark"
        theme_name = self.apply(target_mode, explicit_theme_name)
        return target_mode, theme_name

File: src/core/orchestrator.ts
import { config } from '../config';
import { generateArtifacts } from './artifacts';
import { normalizeMode } from './themeModel';
import { ThemeRegistry } from './themeRegistry';
import { Palette } from '../colors';

const logger = {
  warn: (m: string) => console.warn('[daybreak]', m),
  debug: (m: string) => {},
};

export interface SystemAdapter {
  name: string;
  getCurrentMode(): string;
  setMode(mode: string, palette?: Palette): void;
}

export interface TerminalAdapter {
  name: string;
  applyMode(mode: string, themeName: string, palette?: Palette): void;
}

export class ThemeOrchestrator {
  private systemAdapter: SystemAdapter | null;
  private terminalAdapters: TerminalAdapter[];
  private registry: ThemeRegistry;

  constructor(
    systemAdapter?: SystemAdapter | null,
    terminalAdapters?: TerminalAdapter[],
    registry?: ThemeRegistry,
  ) {
    this.systemAdapter = systemAdapter || null;
    this.terminalAdapters = terminalAdapters || [];
    this.registry = registry || new ThemeRegistry();
  }

  getCurrentMode(): string {
    if (!this.systemAdapter) return 'light';
    try {
      return normalizeMode(this.systemAdapter.getCurrentMode());
    } catch (exc) {
      logger.warn(`Failed to detect current mode. Defaulting to light: ${exc}`);
      return 'light';
    }
  }

  apply(mode: string, themeName?: string): string {
    config.reload();
    const normalizedMode = normalizeMode(mode);
    const resolvedTheme = themeName || config.getModeThemeName(normalizedMode);
    const palette = this.registry.getPalette(resolvedTheme, normalizedMode);

    if (this.systemAdapter) {
      this.systemAdapter.setMode(normalizedMode, palette);
    }

    for (const adapter of this.terminalAdapters) {
      try {
        adapter.applyMode(normalizedMode, resolvedTheme, palette);
      } catch (exc) {
        const name = adapter.name || adapter.constructor.name;
        logger.warn(`Terminal adapter '${name}' failed: ${exc}`);
      }
    }

    try {
      const tokens = this.registry.getTokens(resolvedTheme, normalizedMode);
      const accentTokens = this.registry.getAccentTokens(resolvedTheme, normalizedMode);
      generateArtifacts(config.configDir, resolvedTheme, normalizedMode, tokens, accentTokens, palette);
    } catch (exc) {
      logger.warn(`Failed to generate theme artifacts: ${exc}`);
    }

    return resolvedTheme;
  }

  applyToggle(explicitThemeName?: string): [string, string] {
    const targetMode = this.getCurrentMode() === 'dark' ? 'light' : 'dark';
    const themeName = this.apply(targetMode, explicitThemeName);
    return [targetMode, themeName];
  }
}
